dot/sync/syncer_test.go

```diff
@@ -49,265 +56,122 @@ func TestMain(m *testing.M) {
+	// initialise runtime
+	genState, err := rtstorage.NewTrieState(genTrie) //nolint
 	require.NoError(t, err)
 
-	// request should start from block 5 (best block number + 1)
-	syncer.synced = false
-	_, err = syncer.ProcessBlockData(resp.BlockData)
-	require.True(t, errors.Is(err, chaindb.ErrKeyNotFound))
-}
-
-func TestRemoveIncludedExtrinsics(t *testing.T) {
-	syncer := NewTestSyncer(t, false)
+	rtCfg := &wasmer.Config{}
+	rtCfg.Storage = genState
+	rtCfg.LogLvl = 3
 
-	ext := []byte("nootwashere")
-	tx := &transaction.ValidTransaction{
-		Extrinsic: ext,
-		Validity:  &transaction.Validity{Priority: 1},
+	nodeStorage := runtime.NodeStorage{}
+	if stateSrvc != nil {
+		nodeStorage.BaseDB = stateSrvc.Base
+	} else {
+		nodeStorage.BaseDB, err = utils.SetupDatabase(filepath.Join(testDatadirPath, "offline_storage"), false)
+		require.NoError(t, err)
 	}
+	rtCfg.NodeStorage = nodeStorage
 
-	_, err := syncer.transactionState.(*state.TransactionState).Push(tx)
-	require.NoError(t, err)
-
-	exts := []types.Extrinsic{ext}
-	body, err := types.NewBodyFromExtrinsics(exts)
+	rtCfg.CodeHash, err = cfg.StorageState.LoadCodeHash(nil)
 	require.NoError(t, err)
 
-	bd := &types.BlockData{
-		Body: body,
-	}
-
-	msg := &network.BlockResponseMessage{
-		BlockData: []*types.BlockData{bd},
-	}
-
-	_, err = syncer.ProcessBlockData(msg.BlockData)
+	instance, err := wasmer.NewRuntimeFromGenesis(gen, rtCfg)
 	require.NoError(t, err)
 
-	inQueue := syncer.transactionState.(*state.TransactionState).Pop()
-	require.Nil(t, inQueue, "queue should be empty")
-}
+	cfg.BlockState.StoreRuntime(cfg.BlockState.BestBlockHash(), instance)
 
-func TestHandleBlockResponse_NoBlockData(t *testing.T) {
-	syncer := NewTestSyncer(t, false)
-	_, err := syncer.ProcessBlockData(nil)
-	require.Equal(t, ErrNilBlockData, err)
-}
-
-func TestHandleBlockResponse_BlockData(t *testing.T) {
-	syncer := NewTestSyncer(t, false)
+	cfg.BlockImportHandler = new(syncmocks.MockBlockImportHandler)
+	cfg.BlockImportHandler.(*syncmocks.MockBlockImportHandler).On("HandleBlockImport", mock.AnythingOfType("*types.Block"), mock.AnythingOfType("*storage.TrieState")).Return(func(block *types.Block, ts *rtstorage.TrieState) error {
+		// store updates state trie nodes in database
+		if err = stateSrvc.Storage.StoreTrie(ts, &block.Header); err != nil {
+			logger.Warn("failed to store state trie for imported block", "block", block.Header.Hash(), "error", err)
+			return err
+		}
 
-	rt, err := syncer.blockState.GetRuntime(nil)
-	require.NoError(t, err)
-
-	parent, err := syncer.blockState.(*state.BlockState).BestBlockHeader()
-	require.NoError(t, err)
+		// store block in database
+		if err = stateSrvc.Block.AddBlock(block); err != nil {
+			if err == blocktree.ErrParentNotFound && block.Header.Number.Cmp(big.NewInt(0)) != 0 {
+				return err
+			} else if err == blocktree.ErrBlockExists || block.Header.Number.Cmp(big.NewInt(0)) == 0 {
+				// this is fine
+			} else {
+				return err
+			}
+		}
```

Comment:
err = stateSrvc.Block.AddBlock(block)
		switch {
		case errors.Is(err, blocktree.ErrParentNotFound):
			if block.Header.Number.Cmp(big.NewInt(0)) != 0 {
				return err
			}
		case errors.Is(err, blocktree.ErrBlockExists):
			break
		case block.Header.Number.Cmp(big.NewInt(0)) == 0:
			break
		default:
			return err
		}